src/mosquitto_stubs.c

```diff
@@ -217,132 +267,102 @@ CAMLprim value mqtt_publish(value mqtt, value msg) {
   CAMLparam3(mqtt, topic, qos);
-  CAMLlocal2(perrno, result);
+  CAMLlocal1(result);
 
   char *topic_;
   size_t topic_len;
   struct ocmq *mq;
-  int rc, lerrno, qos_;
+  int rc, qos_;
 
   qos_ = Long_val(qos);
-  mq = (struct ocmq*)mqtt;
+  mq = *(OCMQ(mqtt));
+  if (mq->conn == NULL) caml_invalid_argument("Mosquitto: already destroyed");
 
   topic_len = caml_string_length(topic);
   topic_ = calloc(1, topic_len + 1);
   memcpy(topic_, String_val(topic), topic_len);
   topic_[topic_len] = '\0';
 
   caml_release_runtime_system();
-
   rc = mosquitto_subscribe(mq->conn, NULL, topic_, qos_);
+  int lerrno = errno;
+  caml_acquire_runtime_system();
+  result = wrap_result(rc, Val_unit, lerrno);
   free(topic_);
 
-  caml_acquire_runtime_system();
+  CAMLreturn(result);
+}
 
-  if (MOSQ_ERR_SUCCESS == rc) {
-    result = RESULT_OK;
-    Store_field(result, 0, Val_unit);
-  } else {
-    switch (rc) {
-      case MOSQ_ERR_INVAL: lerrno = EINVAL; break;
-      case MOSQ_ERR_NOMEM: lerrno = ENOMEM; break;
-      case MOSQ_ERR_NO_CONN: lerrno = ENOTCONN; break;
-      default: assert(false);
-    }
-    perrno = caml_alloc(2, 0);
-    Store_field(perrno, 0, eunix); // `EUnix
-    Store_field(perrno, 1, unix_error_of_code(lerrno));
+static value extract_message_tuple(const struct mosquitto_message *msg_) {
+  CAMLparam0();
+  CAMLlocal3(msg, payload, topic);
 
-    result = RESULT_ERROR;
-    Store_field(result, 0, perrno);
-  }
+  size_t topic_len;
 
-  CAMLreturn(result);
+  msg = caml_alloc_tuple(5);
+
+  payload = caml_alloc_initialized_string(msg_->payloadlen, msg_->payload);
+
+  topic_len = strlen(msg_->topic);
+  topic = caml_alloc_initialized_string(topic_len, msg_->topic);
+
+  Store_field(msg, 0, Val_int(msg_->mid));
+  Store_field(msg, 1, topic);
+  Store_field(msg, 2, payload);
+  Store_field(msg, 3, Val_int(msg_->qos));
+  Store_field(msg, 4, Val_bool(msg_->retain));
+
+  CAMLreturn(msg);
 }
 
 void mqtt_callback_msg(struct mosquitto *m, void *obj, const struct mosquitto_message *msg_) {
-  static value * f = NULL;
   struct ocmq *mq;
-  size_t topic_len;
-  value msg, payload;
+  value msg;
 
   mq = (struct ocmq*)obj;
+  if (mq->conn == NULL) return;
 
   caml_acquire_runtime_system();
 
-  if (NULL == f) {
-    f = caml_named_value(mq->uid[CBMESSAGE]);
+  if (NULL == mq->cb[CBMESSAGE]) {
+    mq->cb[CBMESSAGE] = caml_named_value(mq->uid[CBMESSAGE]);
   }
-  if (NULL != f) {
-    msg = caml_alloc_tuple(5);
-
-    payload = caml_alloc_string(msg_->payloadlen);
-    memcpy(String_val(payload), msg_->payload, msg_->payloadlen);
-
-    topic_len = strlen(msg_->topic);
-    value topic = caml_alloc_string(topic_len);
-    memcpy(String_val(topic), msg_->topic, topic_len);
-
-    Store_field(msg, 0, Val_int(msg_->mid));
-    Store_field(msg, 1, topic);
-    Store_field(msg, 2, payload);
-    Store_field(msg, 3, Val_int(msg_->qos));
-    Store_field(msg, 4, Val_bool(msg_->retain));
-
-    caml_callback(*f, msg);
```

Comment:
complying to the rules of the GC by factoring this block out into a helper function so that we can use the `CAMLlocal` macros